ThunkLibs/libX11/libX11_Host.cpp

```diff
@@ -141,29 +166,36 @@ void *libX11_Variadic_u64(uint64_t a_0, size_t count, unsigned long *list, void
       # Store how much data we added to the stack in our callee saved register we stole
       mov x28, x10
 
-      # Subtract one member due to nullptr ender
-      sub x10, x1, 1
-
-      # x11 - stack offset
+      # x11 - stack offset - for stack Arguments
       mov x11, sp
 
-      # x12 - load offset
+      # x12 - load offset - From array source.
+      # Skipping the first seven arguments.
       add x12, x9, (7 * 8)
 
-      cmp x10, 1
+      # x1 currently contains the number of elements including nullptr.
+      # We want the number of arguments to store, not including the nullptr.
+      subs x10, x1, 1
+
       b.eq .single%=
```

Comment:
These paths are just taking advantage of the hardware.
`load_pair` - is just using ldp+stp to copy 128-bits per step.
`single` - uses ldr+stp to load 64-bit and store 128-bit (the final nullptr) in a step.
`no_single` is the case in which there is only the final nullptr to store. So it does it in a single str.

The passed in array doesn't have a nullptr at the end since there is no reason to eat a stack slot in that case.

Reply:
Does this make a measurable difference? Can we use a branchless implementation instead to keep complexity more contained by not taking advantage of the hardware?

Reply:
There's definitely going to be a loop still, so it won't be branchless.
Technically it makes a measurable difference, because ldp+stp has the same performance as ldr+str but it is moving double the number of bits. 64 versus 128.

Reply:
This reads to me like the data suggests it makes no practical difference on performance? That would make the loop probably the more maintainable solution. I don't know if it makes a big difference, but I know I'm unhappy with unjustified complexity in this function.

Reply:
It has exactly 2x the difference in performance.

Reply:
That's referring to the individual instructions, not the overall practical effect. There is little need in micro-optimizing here considering the whole function is wrapped by one heap allocation and potentially more.

Reply:
Removed the optimization. It now just copies one 64-bit value at a time.

Reply:
(Note I'm not sure what the single-branch alternative would look like, so for all I know it wouldn't make all that much difference. I'm just pointing out the current code is overly complex for the context it's used in.)